src/tool/subcommands/api_cmd/generate_test_snapshot.rs

```diff
@@ -0,0 +1,233 @@
+    let network_name = get_network_name_from_genesis(&genesis_header, &state_manager)?;
+    let message_pool = MessagePool::new(
+        MpoolRpcProvider::new(chain_store.publisher().clone(), state_manager.clone()),
+        network_name.clone(),
+        network_send.clone(),
+        Default::default(),
+        state_manager.chain_config().clone(),
+        &mut JoinSet::new(),
+    )?;
+
+    let peer_manager = Arc::new(PeerManager::default());
+    let sync_network_context =
+        SyncNetworkContext::new(network_send, peer_manager, state_manager.blockstore_owned());
+    let (shutdown, shutdown_recv) = mpsc::channel(1);
+    let rpc_state = Arc::new(RPCState {
+        state_manager,
+        keystore: Arc::new(tokio::sync::RwLock::new(KeyStore::new(
+            KeyStoreConfig::Memory,
+        )?)),
+        mpool: Arc::new(message_pool),
+        bad_blocks: Default::default(),
+        sync_state: Arc::new(RwLock::new(Default::default())),
+        eth_event_handler: Arc::new(EthEventHandler::new()),
+        sync_network_context,
+        network_name,
+        start_time: chrono::Utc::now(),
+        shutdown,
+        tipset_send,
+    });
+    rpc_state.sync_state.write().set_stage(SyncStage::Idle);
+    Ok((rpc_state, network_rx, shutdown_recv))
+}
+
+/// A [`Blockstore`] wrapper that tracks read operations to the inner [`Blockstore`] with an [`MemoryDB`]
+pub struct ReadOpsTrackingStore<T> {
+    inner: T,
+    pub tracker: Arc<MemoryDB>,
+}
+
+impl<T> ReadOpsTrackingStore<T> {
+    pub fn new(inner: T) -> Self {
+        Self {
+            inner,
+            tracker: Arc::new(Default::default()),
+        }
+    }
+}
+
+impl<T: Blockstore> Blockstore for ReadOpsTrackingStore<T> {
+    fn get(&self, k: &Cid) -> anyhow::Result<Option<Vec<u8>>> {
+        let result = self.inner.get(k)?;
+        if let Some(v) = &result {
+            self.tracker.put_keyed(k, v.as_slice())?;
+        }
+        Ok(result)
+    }
+
+    fn put_keyed(&self, k: &Cid, block: &[u8]) -> anyhow::Result<()> {
+        self.inner.put_keyed(k, block)
+    }
+}
+
+impl<T: SettingsStore> SettingsStore for ReadOpsTrackingStore<T> {
+    fn read_bin(&self, key: &str) -> anyhow::Result<Option<Vec<u8>>> {
+        let result = self.inner.read_bin(key)?;
+        if let Some(v) = &result {
+            SettingsStore::write_bin(&self.tracker, key, v.as_slice())?;
+        }
+        Ok(result)
+    }
+
+    fn write_bin(&self, key: &str, value: &[u8]) -> anyhow::Result<()> {
+        self.inner.write_bin(key, value)
+    }
+
+    fn exists(&self, key: &str) -> anyhow::Result<bool> {
+        let result = self.inner.read_bin(key)?;
+        if let Some(v) = &result {
+            SettingsStore::write_bin(&self.tracker, key, v.as_slice())?;
+        }
+        Ok(result.is_some())
+    }
+
+    fn setting_keys(&self) -> anyhow::Result<Vec<String>> {
+        // HACKHACK: may need some care
```

Comment:
We could deal with these `HACKHACK` later when they are used by some RPC methods and the snap generation flow fails